### ISSUE TITLE
5.0.1 fails to build on Android?

### ISSUE THREAD

**Issue by gnzlbg** [jemalloc/jemalloc#1175]
Currently `servo` failed to upgrade to jemalloc 5.0.1 due to a build failure on android.

The log is here (read from end): http://build.servo.org/builders/android/builds/11294/steps/compile/logs/stdio

```
jemalloc/src/background_thread.c:83:2: error: unknown type name 'cpu_set_t'
cpu_set_t cpuset;
^
```

Is Android part of jemalloc's CI, or are there any peculiarities that must be taken into account to compile jemalloc 5.x on Android?

Rust is also stuck on jemalloc 4.5 due to this same issue: https://github.com/rust-lang/rust/pull/45163#issuecomment-336491499

**Comment by SimonSapin** [comment]
(I don’t know if the Rust failure were due to the same issue, I can’t find `cpu_set_t` in those logs.)

**Comment by interwq** [comment]
I don't think we have CI on android right now. Are you aware of anything there @davidtgoldblatt?

The configure output says:
`
checking for sched_getcpu... yes

checking for sched_setaffinity... yes
`

However it cannot find sched_getcpu and cpu_set_t. It's worth checking if they require anything additional to sched.h and _GNU_SOURCE.

BTW, --disable-tls is now removed from jemalloc 5.

**Comment by interwq** [comment]
@SimonSapin : feel free to let us know if we can help. 

Regarding the Android CI, I'm afraid at the moment none of the core developers has much experience on the platforms there. If you guys have a relatively straightforward setup (e.g. a list of shell scripts / commands), we'd be happy to integrate it into our CI setups.

**Comment by gnzlbg** [comment]
@interwq 

> If you guys have a relatively straightforward setup (e.g. a list of shell scripts / commands), we'd be happy to integrate it into our CI setups.

This PR shows how we test android on travis: https://github.com/rust-lang-nursery/stdsimd/pull/369

* this file gets/installs the android-ndk on linux: https://github.com/rust-lang-nursery/stdsimd/pull/369/files#diff-6fbc6d292572e541f5bc416175bef165

* this one gets/installs the android-sdk: https://github.com/rust-lang-nursery/stdsimd/pull/369/files#diff-91491aa6437f307539dff754ba211aad

* this one a sysimage: https://github.com/rust-lang-nursery/stdsimd/pull/369/files#diff-6d774679c1d2558be216c1bf8965fd1e

* then we have a bunch of docker containers for different android targets:

    * aarch64: https://github.com/rust-lang-nursery/stdsimd/pull/369/files#diff-0d1ad4513509be2ac2b23c86b7050db8
    * x86_64: https://github.com/rust-lang-nursery/stdsimd/pull/369/files#diff-9ef6f5bd0ed37d42490a496a883daa94 
    * arm: https://github.com/rust-lang-nursery/stdsimd/pull/369/files#diff-bcd351ab73457cabc0bb37469b02b7c4

* and finally one little program that takes as input a binary to run, initializes the android emulator (or detect whether one has been initialized), copies the binary to the device, executes, and fetches stdout,stderr, prints them, and fetches the return code, and returns it: https://github.com/rust-lang-nursery/stdsimd/pull/369/files#diff-ff8a1e27a86f7f5a6ccad985336919ed

You can then use this program as a "runner" for your tests inside the docker containers.

I don't know how/where you are testing jemalloc, but if you are using travis-ci or linux it should be pretty straight-forward to integrate something like this. If you point me to your CI system and can take a look, or otherwise feel free to ask any questions.

**Comment by interwq** [comment]
Thanks a lot for sharing the detailed information @gnzlbg! We use Travis and AppVeyor, combined with our customized CIs. We'll try this out (however may take a while before we get someone working on it).

**Comment by gnzlbg** [comment]
@interwq if you are using travis then building CI for android in is doable.

Let me know if I can help.

---

FYI [these](https://travis-ci.org/rust-lang/libc) are the platforms that Rust's libc tests with a similar setup. If there are some of them that are not covered by your CI and want help setting any of those up let me know as well. Most of them are tier2 (we guarantee cross-compiling to them suceed, but don't guarantee that the binaries run properly). It would help those interested into making them tier1 platforms to know that the memory allocator at least upstream works just fine there.

**Comment by gnzlbg** [comment]
@interwq is `_GNU_SOURCE` defined when including `sched.h` on android ? 

This SO post (https://stackoverflow.com/a/16323804/1422197) suggest, as a workaround, to define them manually:

```c++
#define CPU_SETSIZE 1024
#define __NCPUBITS  (8 * sizeof (unsigned long))
typedef struct
{
   unsigned long __bits[CPU_SETSIZE / __NCPUBITS];
} cpu_set_t;

#define CPU_SET(cpu, cpusetp) \
  ((cpusetp)->__bits[(cpu)/__NCPUBITS] |= (1UL << ((cpu) % __NCPUBITS)))
#define CPU_ZERO(cpusetp) \
  memset((cpusetp), 0, sizeof(cpu_set_t))
```

but that will collide with the definitions exported by `sched.h` in other platforms (or later versions of the Android NDK or the API version).

**Comment by jdm** [comment]
To close the loop, right now Servo relies on NDK 12 and API level 18.

**Comment by interwq** [comment]
@gnzlbg: we should have -D_GNU_SOURCE in the compile flags. Can you please check if that's not included somehow for your case?

**Comment by gnzlbg** [comment]
@interwq that's indeed the case for Rust upstream on android. I am going to add a buildbot for the NDK 12 and API level 18 and see if I can reproduce this with Rust upstream. On later android versions we get currently run into #1203 on arm-android and armv7-android but not aarch64-android.

**Comment by gnzlbg** [comment]
@interwq So I've tried disabling the background thread, but jemalloc still fails to build on that android version: http://build.servo.org/builders/android/builds/11639/steps/compile/logs/stdio

For some reason, even though it appears that the background thread is disabled properly:

> malloc_conf        : background_thread:false

jemalloc still builds `background_thread.c`. Also, there is a bunch of threading related stuff in `jemalloc_internal_inlines_a.h` that doesn't seem to be `#ifdef`ed out when the background thread is disabled.
